### ISSUE TITLE
Proposal for a generic filter option for any list

### ISSUE THREAD

**Issue by agrogers** [AndBible/and-bible#1620]
My lists are starting to get very long. I am usually scrolling for seconds to get to the item I want. This applies to Study Pads, Labels, Dictionary entries, bookmarks, label settings etc

What I was originally thinking of was a simple alphabetical chooser. It shows 'A B C D E F etc' in a row at the top which would filter my lists by first letter. This would get me to most of the things i want with one click and no scrolling. But I have no idea how to implement that for other languages. And it is somewhat wasteful to have, say, X and Z in the list which are rarely used.

The proposal below would work for all languages and be easily customisable:

- A search icon is shown at the top of the list we want to filter
- Pushing it shows a search text box. 
- The text box works like existing filter text boxes - It immediately shows the filtered list below. The difference is...
- Whatever the user types gets saved in the recently searched list. For me i would just type one letter - 'G' say. Then when I return to the form the letter 'G' will appear in the list of search terms. I could, if i wanted, add all the letters in just a minute or so. Then, I can easily filter on the first letter. This should work for any language.
- But i can also add words or multiple letter groups.
- Long pressing the filter option would remove that option from the list. So it would be easy to keep the list clean
- I have shown a single line of saved filter options. But it could be two rows or more deep.
- The same or custom saved lists could be used on all lists that need filtering
- I would also be tempted to remember what was searched for last time and the filtered list shown when returning to it. If that was done we might need an X option to clear the filter.

![image](https://user-images.githubusercontent.com/13920678/130425212-4feda342-ca3b-4592-8657-9af06e74954f.png)

**Comment by Dleemuse** [comment]
Love, love, love this idea! I have so many bookmarks, etc., this would be a time saver.

Debra

**Comment by dbell154** [comment]
If you are just talking about scrolling, why not use the fast scroll setting in android. You can drag the scrollbar thumb thing to the bottom of the list with one stroke.
I like the search option also.
On a similar note, I think it would be nice to have a way to assign labels to different label groups.

**Comment by agrogers** [comment]
How would label groups work?

**Comment by timbze** [comment]
Something like #1025 may be what @dbell154 is asking for

**Comment by agrogers** [comment]
Ah, i get it.

___________________________
*www.TenOutOfTen.org* <https://www.TenOutOfTen.org>
***@***.***
(+61) 0477 174 193 (Australia)
(+95) 09 250018669 (Myanmar)

**Comment by dbell154** [comment]
What about fast scrolling like what is in the Contacts screen (mine anyway)? Letters are shown on the right and when you fast scroll the letter for at the current position is displayed. Similar to the exercise at http://androidopentutorials.com/android-listview-fastscroll/.

**Comment by agrogers** [comment]
Letters on the right would also be helpful. Although I would also like to be able to filter by any part of a label. I need to try and do something soon though. I spend a lot of time scrolling through my ever increasing label list :)

**Comment by dbell154** [comment]
One way to do this would be to use phases.
1. Enable fast scroll on the listview. This can be done in xml or code. Very easy to do.
2. Add a setting that lets users choose to enable fast scroll, and enable in code based on the setting.
3. The letters on the side would be a cool enhancement, though the filter (next) may be more useful.
4. Add a filter option for the list instead of search. There's a javascript function I used in one of my web pages that would filter the page on a by table row basis. (http://www.tylerhosting.com/voltax/taxslayer/index.html). As you type, the list limits what is displayed. Shouldn't be too hard to implement with the database. Select label where label like % (entry) %.
5. Add a table to record searches as you illustrated, although I worry that a list of past searches might get too unwieldy pretty quick.
I've been trying to figure out the "scroll with letters" implementation but with limited Kotlin knowledge and some distractions I haven't gotten very far.

**Comment by agrogers** [comment]
I have mostly implemented this now and got a draft PR here https://github.com/AndBible/and-bible/pull/2058

1. Need to look into fast scrolling. It can take a long time for me now.
2. Letters down the side could be good too. Although i find I have too little width for some long labels so i think the top would be better for me
3. The filter option has been added now - filters while you type
4. There is save option to save the filter. I implemented this mainly because I didn't know how to create a list of filter buttons for every language we have available. I am hoping this approach allows people to create their own language specific list. Not sure how the right to left  languages will go though.
5. The search uses standard regex 
6. The same saved lists apply to all lists
7. Additional suggestions are welcome
8. Oh, it also always shows the selected labels regardless of filter. This seems right to me.

![image](https://user-images.githubusercontent.com/13920678/150125798-2c854095-7560-470f-9816-af034ab3a595.png)

**Comment by dbell154** [comment]
Looks good.
I started working on this but it didn't take long before I was over my head. The one thing that I did do was add fast scrolling which only consists of two lines under ListView in the xml file manage_labels.xml.

		android:fastScrollAlwaysVisible="true"
		android:fastScrollEnabled="true"

I use fast scroll in my word study app. The touch area for fast scroll is a little wider than I like so I have an overlay at the left edge and users must enable it in settings. I don't think that would be necessary in this case.

With the buttons at the top and fast scroll enabled I don't think the side letters would be necessary.

**Comment by agrogers** [comment]
Cool. I will check it out.

**Comment by agrogers** [comment]
I have added the fast scroll in now Dana. Can't seem to get the thumb image thing popping up. 

I have also added some code so that it recognises text like 'A-C' as all the labels between A and C. 

And I fixed a bug so that it is possible to enter any valid REGEX. So "^(?!God)G" what find all G words excluding those that start  with 'God'

![image](https://user-images.githubusercontent.com/13920678/150242356-d6dc12a0-210c-4f35-85f8-2bbda209949a.png)
